tests/5.2/declare_mapper/test_declare_mapper_present.c

```diff
@@ -0,0 +1,67 @@
+//===---- test_declare_mapper_present.c -------------------------------------===//
+// 
+// OpenMP API Version 5.2 
+//
+// This example has been adapted from the 5.2 OpenMP Examples document:
+// "declare mapper Directive", and "OpenMP Directive Syntax -> complex iterator"
+// The declare mapper directive will be used to automatically map variables
+// according to its prescription:  full structure, plus the dynamic storage of the
+// data element. These variable should be present on the device.
+//
+//===-------------------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1000
+
+typedef struct myvec{
+    size_t len;
+    int *data;
+} myvec_t;
+
+
+
+void init( myvec_t *s )
+{ 
+  for(int i = 0; i < s->len; i++)
+    s->data[i] = i; 
+}
+
+
+int test_declare_mapper_present() { 
+
+  OMPVV_INFOMSG("test_declare_mapper_present");
+  int errors = 0;
+
+  myvec_t s;
+
+  s.data = (int *)calloc(N,sizeof(int));
+  s.len  = N;
+  #pragma omp target enter data map(to: s, s.data[0:s.len])
+
+  #pragma omp declare mapper(default:myvec_t v) map(present, from: v, v.data[0:v.len]) 
+
+  #pragma omp target map(s)
```

Comment:
Do we need map(s) here ?

Reply:
map(s) does the same, without a explicit modifier it is "tofrom".

Reply:
But a bare `map(tofrom: s)` won't help here, either – I mean because it is already present?!?

Thus, before testing, IMHO
* either an "**always**" modifier is needed in the mapper or the usage of the mapper. _(I think it is permitted, but I vaguely remember some more recent changes; either in 5.2 or later, not checked.)_ 
* Or after the target region, the ref count has to be reduced to force the copy out (like: **target exit data**)
* or the data has to be explicitly be copied (i.e. using the **target update** directive).

Reply:
OpenMP V5.1 Section 2.21.7.1 map Clause says:

> If the mapper map-type-modifier is not present, the behavior is as if the mapper(default) modifier was specified. 

This means that the `mapper(default)` will be applied to the `map` clause by default.

Reply:
@tob2 I agree that this test will fail because it is already present.
(Test in line 53 will fail as commented below.)